src/java/org/apache/cassandra/db/commitlog/CommitLogSegmentReader.java

```diff
@@ -101,56 +101,69 @@ protected SyncSegment computeNext()
         {
             while (true)
             {
+                final int currentStart = end;
                 try
                 {
-                    final int currentStart = end;
                     end = readSyncMarker(descriptor, currentStart, reader);
-                    if (end == -1)
-                    {
-                        return endOfData();
-                    }
-                    if (end > reader.length())
-                    {
-                        // the CRC was good (meaning it was good when it was written and still looks legit), but the file is truncated now.
-                        // try to grab and use as much of the file as possible, which might be nothing if the end of the file truly is corrupt
-                        end = (int) reader.length();
-                    }
+                }
+                catch (CommitLogSegmentReader.SegmentReadException e)
+                {
+                    handleUnrecoverableError(e, !e.invalidCrc && tolerateTruncation);
+                    end = -1; // skip the remaining part of the corrupted log segment
```

Comment:
So just to iterate on the logic here ... when `readSyncMarker` throws `SegmentReadException` and it is evaluated as it is _permissible_, then `handleUnrecoverableError` will not throw, so `end` will be set to `-1`, hence we will `return endOfData();` later on.

If it is thrown and not permissible, we will throw `RuntimeException` from `handleUnrecoverableError`. 

Then I see that when it is thrown like that, reader is iterated over in `CommitLogReader#readCommitLogSegment`, which is wrapped by try-catch and catches RTE and throws again ... 

Yeah I think this is legit.

Reply:
yes, the essential change is adding end = -1 to move to the end of the corrupted segment, previously when we ignored an error we started to read the corrupted place again from the same place, so it was an infinite loop. 
handleUnrecoverableError contains the existing logic, I've just moved the repeating logic into an utility method to make the main loop logic more clear.